morphman/manipulate_branch.py

```diff
@@ -0,0 +1,619 @@
+
+    # Clip Voronoi diagram into bend and remaining part of geometry
+    print("-- Clipping Voronoi diagrams")
+    voronoi_branch, voronoi_remaining = get_split_voronoi_diagram(voronoi, [diverging_centerline_branch,
+                                                                            centerlines])
+
+    # Get surface normals
+    old_normal = get_estimated_surface_normal(diverging_centerline_branch)
+    new_normal = get_exact_surface_normal(capped_surface, new_branch_pos_id)
+
+    # Define rotation between surface normal vectors
+    u, surface_normals_angle = get_rotation_axis_and_angle(new_normal, old_normal)
+    R_u = get_rotation_matrix(u, surface_normals_angle)
+
+    # Define rotation around new surface normal
+    R_z = get_rotation_matrix(-new_normal, angle)
+
+    # Define translation parameters
+    dx, origo = get_translation_parameters(centerlines, diverging_centerline_branch, new_branch_pos)
+
+    # Move branch centerline and voronoi diagram
+    moved_and_rotated_voronoi_branch = move_and_rotate_voronoi_branch(voronoi_branch, dx, R_u, R_z, origo)
+    moved_and_rotated_centerline_branch = move_and_rotate_centerline_branch(diverging_centerline_branch, origo, R_u,
+                                                                            R_z, dx)
+    # Create new voronoi diagram and new centerlines
+    new_voronoi = vtk_merge_polydata([voronoi_remaining, moved_and_rotated_voronoi_branch])
+    write_polydata(new_voronoi, new_voronoi_path)
+
+    new_centerlines = vtk_merge_polydata([centerlines, moved_and_rotated_centerline_branch])
+    write_polydata(new_centerlines, new_centerlines_path)
+
+    new_surface = create_new_surface(new_voronoi, poly_ball_size=poly_ball_size)
+    write_polydata(new_surface, unprepared_output_filepath)
+
+    print("-- Smoothing, clean, and check surface")
+    new_surface = prepare_output_surface(new_surface, surface,
+                                         new_centerlines, output_filepath,
+                                         test_merge=False, changed=True,
+                                         old_centerline=centerlines_complete)
+
+    write_polydata(new_surface, output_filepath)
+
+
+def check_branch_number(branch_to_manipulate_number, centerlines_complete):
+    """
+    Check if branch number provided by user is larger than number of centerlines.
+    Rises RuntimeError if number exceeds limit.
+
+    Args:
+        branch_to_manipulate_number (int): Input number, supplied by user
+        centerlines_complete (vtkPolyData): All centerlines
+    """
+    num_lines = centerlines_complete.GetNumberOfLines() - 1
+    if branch_to_manipulate_number > num_lines:
+        raise RuntimeError("\nERROR: Branch number cannot exceed number of centerlines." +
+                           " Number of selectable centerlines for this model is {}.".format(num_lines))
+
+
+def vmtk_compute_surface_normals(capped_surface):
```

Comment:
I would prefer if the vmtk functions are in vmtk_wrapper.py in common. In fact I think we compute the normals at least one other location in the code. This could be replaced by this function,

Reply:
Good point! I'll look for duplicate code, but the placement of these local vmtk functions were due to external imports from morphman (`from morphman.common.vmtk-wrapper...`) was missing, thus causing import errors when running locally. However, the build/CI should pass, so I'll try extracting these vmtk functions in a later commit 👌